src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala

```diff
@@ -0,0 +1,663 @@
+    if (readStructures.length != fastqs.count(_.nonEmpty)) throw new IllegalArgumentException("The # of read structures did not match the number of fastqs.")
+
+    // check the molecular barcodes and associated tags, if any exist.
+    val molecularBarcodes = readStructures.flatMap { rs => rs.molecularBarcode }
+    if (molecularBarcodes.isEmpty && multiUmiTags.nonEmpty) {
+      throw new IllegalArgumentException("No molecular barcodes found but molecular barcode SAM tags were given.")
+    }
+    if (multiUmiTags.length > 1 && multiUmiTags.length != molecularBarcodes.length) {
+      throw new IllegalArgumentException("Either one SAM tag or the same number as the molecular barcode segments should be given " +
+        s"(found ${molecularBarcodes.length} molecular barcodes but given ${multiUmiTags.length} tags."
+      )
+    }
+
+    // Read the sample sheet, convert to a custom sample type, and set the barcode bases for each sample
+    val sampleSheet: SampleSheet = DemuxFastqs.sampleSheet(sampleSheet=this.sampleSheet, lane=lane).sampleBarcodes(readStructures:_*)
+
+    // If the sample barcodes are in-line in the read (i.e. in r1 or r2) then they may be reverse complemented (ex. amplification).
+    if (matchReverseComplement) {
+      // check that we only have in-line sample barcodes.
+      if (Seq(i7ReadStructure, i5ReadStructure).exists(_.exists(_.sampleBarcode.nonEmpty))) {
+        throw new IllegalArgumentException("Cannot match the reverse complement of the sample barcode when sample barcodes are in the index reads.")
+      }
+      // update the sample barcode to include the reverse complement
+      sampleSheet.foreach { sample =>
+        sample.sampleBarcode match {
+          case Some(sampleBarcode) => sample.sampleBarcode = Some(new CustomSampleBarcode(Seq(new String(sampleBarcode.barcodeBytes))))
+          case None => unreachable("sample barcode should have been set.")
+        }
+      }
+      // NB: we could check that no sample barcodes are reverse complements of each other.
+    }
+
+    // Determine the quality format
+    determineQualityFormat()
+
+    // Create the files for writing
+    val writers = createSamFileWriters(sampleSheet)
+
+    // Open the input FASTQs for reading
+    val nonEmptyFastqs = fastqs.filter(_.nonEmpty)
+    val reader = new ParallelFastqReader(nonEmptyFastqs.head, nonEmptyFastqs.tail:_*)
+
+    // Read in the quadruples of FASTQ records
+    val fastqConverter = new FastqConverter(
+      reader,
+      writers.map(_.getFileHeader),
+      sampleSheet,
+      umiTag,
+      multiUmiTags,
+      qualityFormat.getOrElse(unreachable("Quality format not set")),
+      minQ,
+      maxQ,
+      maxMismatches,
+      minMismatchDelta,
+      maxNoCalls,
+      matchReverseComplement,
+      readStructures:_*)
+
+    val progress = new com.fulcrumgenomics.util.ProgressLogger(logger)
```

Comment:
Why the full package name?  Can't you just import it?  Or import it as `=> FgProgressLogger` if you need to?